src/index.cpp

```diff
@@ -1485,12 +1706,123 @@ namespace diskann {
+  void Index<T, TagT, LabelT>::parse_label_file(const std::string &map_file) {
+    // Format of Label txt file: filters with comma separators
+
+    std::ifstream infile(map_file);
+    std::string   line, token;
+    unsigned      line_cnt = 0;
+
+    while (std::getline(infile, line)) {
+      line_cnt++;
+    }
+    _pts_to_labels.resize(line_cnt, std::vector<LabelT>());
+
+    infile.clear();
+    infile.seekg(0, std::ios::beg);
+    line_cnt = 0;
+    while (std::getline(infile, line)) {
+      std::istringstream iss(line);
+      std::vector<LabelT> lbls(0);
+      getline(iss, token, '\t');
+      std::istringstream new_iss(token);
+      while (getline(new_iss, token, ',')) {
+        token.erase(std::remove(token.begin(), token.end(), '\n'), token.end());
+        token.erase(std::remove(token.begin(), token.end(), '\r'), token.end());
+        LabelT token_as_num = std::stoul(token);
+        lbls.push_back(token_as_num);
+        _labels.insert(token_as_num);
+      }
+      if (lbls.size() <= 0) {
+        std::cout << "No label found";
+        exit(-1);
+      }
+      std::sort(lbls.begin(), lbls.end());
+      _pts_to_labels[line_cnt] = lbls;
+      line_cnt++;
+    }
+    std::cout << "Identified " << _labels.size() << " distinct label(s)"
+              << std::endl;
+  }
+
+  template<typename T, typename TagT, typename LabelT>
+  void Index<T, TagT, LabelT>::set_universal_label(const LabelT &label) {
+    _use_universal_label = true;
+    _universal_label = label;
+  }
+
+  template<typename T, typename TagT, typename LabelT>
+  void Index<T, TagT, LabelT>::build_filtered_index(const char *       filename,
+                                            const std::string &label_file,
+                                            const size_t num_points_to_load,
+                                            Parameters & parameters,
+                                            const std::vector<TagT> &tags) {
+    _labels_file = label_file;
+    _filtered_index = true;
+    _filter_to_medoid_id.clear();
+    parse_label_file(label_file);  // determines medoid for each label and
+                                   // identifies the points to label mapping
+
+    _u32 counter = 0;
+#pragma omp parallel for schedule(dynamic, 1)
+    for (int lbl = 0; lbl < _labels.size(); lbl++) {
+      auto itr = _labels.begin();
+      std::advance(itr, lbl);
+      auto &            x = *itr;
+      std::vector<_u32> filtered_points;
+      for (_u32 i = 0; i < num_points_to_load; i++) {
+        if (std::find(_pts_to_labels[i].begin(), _pts_to_labels[i].end(), x) !=
+                _pts_to_labels[i].end() ||
+            (_use_universal_label &&
+             (std::find(_pts_to_labels[i].begin(), _pts_to_labels[i].end(),
+                        _universal_label) != _pts_to_labels[i].end())))
+          filtered_points.emplace_back(i);
+      }
+      if (filtered_points.size() != 0) {
+#pragma omp critical
```

Comment:
Avoid using omp critical. Can this be written another way?

Reply:
Fixed